utilities/integrator.py

```diff
@@ -0,0 +1,118 @@
+import torch
+
+
+class Integrator:
+    """HGN integrator class: Implements different integration methods for Hamiltonian differential equations.
+    """
+    METHODS = ["Euler", "RK4", "leapfrog"]
+
+    def __init__(self, delta_t, method="Euler"):
+        """Initialize HGN integrator.
+
+        Args:
+            delta_t (float): Time difference between integration steps.
+            method (str, optional): Integration method, must be "Euler", "RK4", or "leapfrog". Defaults to "Euler".
+
+        Raises:
+            KeyError: If the integration method passed is invalid.
+        """
+        if method not in self.METHODS:
+            msg = "%s is not a supported method. " % (method)
+            msg += "Available methods are: " + "".join("%s " % m for m in self.METHODS)
+            raise KeyError(msg)
+
+        self.delta_t = delta_t
+        self.method = method
+
+    def _get_grads(self, q, p, hnn):
+        """Apply the Hamiltonian equations to the Hamiltonian network to get dq_dt, dp_dt.
+
+        Args:
+            q (torch.Tensor): Latent-space position tensor.
+            p (torch.Tensor): Latent-space momentum tensor.
+            hnn (HamiltonianNet): Hamiltonian Neural Network.
+
+        Returns:
+            tuple(torch.Tensor, torch.Tensor): Position and momentum time derivatives: dq_dt, dp_dt.
+        """
+        # Compute energy of the system
+        energy = hnn(q=q, p=p)
+
+        # dq_dt = dH/dp
+        dq_dt = torch.autograd.grad(energy, p, create_graph=True, retain_graph=True)[0]
+        # dp_dt = -dH/dq
+        dp_dt = -torch.autograd.grad(energy, q, create_graph=True, retain_graph=True)[0]
```

Comment:
Change this as we discussed for allowing minibatches
```suggestion
        dq_dt = torch.autograd.grad(energy, p, create_graph=True, retain_graph=True,
            grad_outputs=torch.ones_like(energy))[0]
        # dp_dt = -dH/dq
        dp_dt = -torch.autograd.grad(energy, q, create_graph=True, retain_graph=True,
            grad_outputs=torch.ones_like(energy))[0]
```